File: code_examples/Algorithms/RadixSort.js
const arr = [303,2923,176,412,609,588,767,9,823,10, 672, 686, 90, 16, 53840, 234, 565]
const getDigit = (num, i) => Math.floor(Math.abs(num) / Math.pow(10, i)) % 10

const numberCount = num => {
    if (num === 0) return 1
    return Math.floor(Math.log10(Math.abs(num))) + 1
}

const mostNumbers = nums => {
    let maxNumbers = 0
    for (let i = 0; i < nums.length; i++) maxNumbers = Math.max(maxNumbers, numberCount(nums[i]))
    return maxNumbers
}

const radixSort = nums => {
    let maxNumberCount = mostNumbers(nums)
    for (let a = 0; a < maxNumberCount; a++) {
        let numberBuckets = Array.from({ length: 10 }, () => [])
        for (let i = 0; i < nums.length; i++) {
            let number = getDigit(nums[i], a)
            numberBuckets[number].push(nums[i])
        }
        nums = [].concat(...numberBuckets)
    }
    return nums
}

console.log(radixSort(arr))

File: code_examples/Algorithms/radixSort.kt
import kotlin.math.abs
import kotlin.math.floor
import kotlin.math.log10
import kotlin.math.pow

fun getDigit(num: Int, i: Int): Int {
    return (floor(abs(num.toDouble()) / 10.0.pow(i.toDouble())) % 10).toInt()
}

fun numberCount(num: Int): Int {
    if (num == 0) return 1
    return (floor(log10(abs(num.toDouble()))) + 1).toInt()
}

fun mostNumbers(nums: List<Int>): Int {
    var maxNumbers = 0
    for (num in nums) {
        maxNumbers = maxOf(maxNumbers, numberCount(num))
    }
    return maxNumbers
}

fun radixSort(nums: MutableList<Int>): List<Int> {
    var numbers = nums.toList()
    val maxNumberCount = mostNumbers(numbers)

    for (a in 0 until maxNumberCount) {
        val numberBuckets = List(10) { mutableListOf<Int>() }
        for (i in numbers.indices) {
            val num = getDigit(numbers[i], a)
            numberBuckets[num].add(numbers[i])
        }
        numbers = numberBuckets.flatten()
    }
    return numbers
}

fun main() {
    val arr = listOf(303, 2923, 176, 412, 609, 588, 767, 9, 823, 10, 672, 686, 90, 16, 53840, 234, 565)
    val sortedArr = radixSort(arr.toMutableList())
    println(sortedArr)
}
